File: pkg/manifest/builder.go
package manifest

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/liv-format/liv/pkg/core"
)

// ManifestBuilder helps create and populate manifest structures
type ManifestBuilder struct {
	manifest  *core.Manifest
	validator *ManifestValidator
}

// NewManifestBuilder creates a new manifest builder
func NewManifestBuilder() *ManifestBuilder {
	return &ManifestBuilder{
		manifest: &core.Manifest{
			Version:   "1.0",
			Resources: make(map[string]*core.Resource),
		},
		validator: NewManifestValidator(),
	}
}

// SetMetadata sets the document metadata
func (mb *ManifestBuilder) SetMetadata(metadata *core.DocumentMetadata) *ManifestBuilder {
	mb.manifest.Metadata = metadata
	return mb
}

// SetSecurityPolicy sets the security policy
func (mb *ManifestBuilder) SetSecurityPolicy(policy *core.SecurityPolicy) *ManifestBuilder {
	mb.manifest.Security = policy
	return mb
}

// SetWASMConfig sets the WASM configuration
func (mb *ManifestBuilder) SetWASMConfig(config *core.WASMConfiguration) *ManifestBuilder {
	mb.manifest.WASMConfig = config
	return mb
}

// SetFeatureFlags sets the feature flags
func (mb *ManifestBuilder) SetFeatureFlags(features *core.FeatureFlags) *ManifestBuilder {
	mb.manifest.Features = features
	return mb
}

// AddResource adds a resource to the manifest
func (mb *ManifestBuilder) AddResource(path string, resource *core.Resource) *ManifestBuilder {
	if mb.manifest.Resources == nil {
		mb.manifest.Resources = make(map[string]*core.Resource)
	}
	mb.manifest.Resources[path] = resource
	return mb
}

// AddResourceFromFile adds a resource by reading from a file
func (mb *ManifestBuilder) AddResourceFromFile(path, filePath string) error {
	file, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("failed to open file %s: %v", filePath, err)
	}
	defer file.Close()

	// Get file info
	info, err := file.Stat()
	if err != nil {
		return fmt.Errorf("failed to get file info for %s: %v", filePath, err)
	}

	// Calculate hash
	hasher := sha256.New()
	if _, err := io.Copy(hasher, file); err != nil {
		return fmt.Errorf("failed to calculate hash for %s: %v", filePath, err)
	}
	hash := hex.EncodeToString(hasher.Sum(nil))

	// Determine MIME type
	mimeType := mb.getMimeType(filePath)

	// Create resource
	resource := &core.Resource{
		Hash: hash,
		Size: info.Size(),
		Type: mimeType,
		Path: path,
	}

	mb.AddResource(path, resource)
	return nil
}

// ScanDirectory scans a directory and adds all files as resources
func (mb *ManifestBuilder) ScanDirectory(baseDir string) error {
	return filepath.Walk(baseDir, func(filePath string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip directories
		if info.IsDir() {
			return nil
		}

		// Calculate relative path
		relPath, err := filepath.Rel(baseDir, filePath)
		if err != nil {
			return fmt.Errorf("failed to get relative path: %v", err)
		}

		// Normalize path separators for cross-platform compatibility
		relPath = filepath.ToSlash(relPath)

		// Add resource
		return mb.AddResourceFromFile(relPath, filePath)
	})
}

// CreateDefaultMetadata creates default metadata with current timestamp
func (mb *ManifestBuilder) CreateDefaultMetadata(title, author string) *ManifestBuilder {
	now := time.Now()
	metadata := &core.DocumentMetadata{
		Title:       title,
		Author:      author,
		Created:     now,
		Modified:    now,
		Description: "",
		Version:     "1.0.0",
		Language:    "en",
	}
	return mb.SetMetadata(metadata)
}

// CreateDefaultSecurityPolicy creates a default security policy
func (mb *ManifestBuilder) CreateDefaultSecurityPolicy() *ManifestBuilder {
	policy := &core.SecurityPolicy{
		WASMPermissions: &core.WASMPermissions{
			MemoryLimit:     64 * 1024 * 1024, // 64MB
			AllowedImports:  []string{},
			CPUTimeLimit:    5000, // 5 seconds
			AllowNetworking: false,
			AllowFileSystem: false,
		},
		JSPermissions: &core.JSPermissions{
			ExecutionMode: "sandboxed",
			AllowedAPIs:   []string{},
			DOMAccess:     "read",
		},
		NetworkPolicy: &core.NetworkPolicy{
			AllowOutbound: false,
			AllowedHosts:  []string{},
			AllowedPorts:  []int{},
		},
		StoragePolicy: &core.StoragePolicy{
			AllowLocalStorage:   false,
			AllowSessionStorage: false,
			AllowIndexedDB:      false,
			AllowCookies:        false,
		},
		ContentSecurityPolicy: "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline';",
		TrustedDomains:        []string{},
	}
	return mb.SetSecurityPolicy(policy)
}

// CreateDefaultFeatureFlags creates default feature flags
func (mb *ManifestBuilder) CreateDefaultFeatureFlags() *ManifestBuilder {
	features := &core.FeatureFlags{
		Animations:    true,
		Interactivity: true,
		Charts:        true,
		Forms:         false,
		Audio:         false,
		Video:         false,
		WebGL:         false,
		WebAssembly:   true,
	}
	return mb.SetFeatureFlags(features)
}

// AddWASMModule adds a WASM module to the configuration
func (mb *ManifestBuilder) AddWASMModule(module *core.WASMModule) *ManifestBuilder {
	if mb.manifest.WASMConfig == nil {
		mb.manifest.WASMConfig = &core.WASMConfiguration{
			Modules: make(map[string]*core.WASMModule),
			Permissions: &core.WASMPermissions{
				MemoryLimit:     128 * 1024 * 1024, // 128MB
				AllowedImports:  []string{},
				CPUTimeLimit:    10000, // 10 seconds
				AllowNetworking: false,
				AllowFileSystem: false,
			},
			MemoryLimit: 128 * 1024 * 1024, // 128MB
		}
	}
	
	mb.manifest.WASMConfig.Modules[module.Name] = module
	return mb
}

// LoadFromFile loads an existing manifest from a JSON file
func (mb *ManifestBuilder) LoadFromFile(filePath string) error {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return fmt.Errorf("failed to read manifest file: %v", err)
	}

	manifest, result := mb.validator.ValidateManifestJSON(data)
	if !result.IsValid {
		return fmt.Errorf("invalid manifest: %v", result.Errors)
	}

	mb.manifest = manifest
	return nil
}

// Build validates and returns the completed manifest
func (mb *ManifestBuilder) Build() (*core.Manifest, error) {
	// Validate the manifest
	result := mb.validator.ValidateManifest(mb.manifest)
	if !result.IsValid {
		return nil, fmt.Errorf("manifest validation failed: %v", result.Errors)
	}

	// Return a copy to prevent further modifications
	manifestCopy := *mb.manifest
	return &manifestCopy, nil
}

// BuildJSON validates and returns the manifest as JSON bytes
func (mb *ManifestBuilder) BuildJSON() ([]byte, error) {
	manifest, err := mb.Build()
	if err != nil {
		return nil, err
	}

	data, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("failed to marshal manifest to JSON: %v", err)
	}

	return data, nil
}

// SaveToFile saves the manifest to a JSON file
func (mb *ManifestBuilder) SaveToFile(filePath string) error {
	// If the manifest directory contains files (content/, assets/, etc.), scan it and add resources
	dir := filepath.Dir(filePath)
	// Scan directory if it exists
	if _, err := os.Stat(dir); err == nil {
		_ = mb.ScanDirectory(dir) // best-effort: add any files found
	}

	// Validate the manifest before serialization
	if result := mb.validator.ValidateManifest(mb.manifest); !result.IsValid {
		return fmt.Errorf("manifest validation failed: %v", result.Errors)
	}

	// Marshal manifest to JSON
	manifest, err := mb.Build()
	if err != nil {
		return err
	}

	data, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal manifest to JSON: %v", err)
	}

	// Create directory if it doesn't exist
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create directory: %v", err)
	}

	// Write manifest file
	if err := os.WriteFile(filePath, data, 0644); err != nil {
		return fmt.Errorf("failed to write manifest file: %v", err)
	}

	return nil
}

// GetManifest returns the current manifest (for inspection)
func (mb *ManifestBuilder) GetManifest() *core.Manifest {
	return mb.manifest
}

// Validate validates the current manifest
func (mb *ManifestBuilder) Validate() *core.ValidationResult {
	return mb.validator.ValidateManifest(mb.manifest)
}

// Helper methods

func (mb *ManifestBuilder) getMimeType(filePath string) string {
	ext := strings.ToLower(filepath.Ext(filePath))
	
	mimeTypes := map[string]string{
		".html": "text/html",
		".htm":  "text/html",
		".css":  "text/css",
		".js":   "application/javascript",
		".json": "application/json",
		".xml":  "application/xml",
		".txt":  "text/plain",
		".md":   "text/markdown",
		
		// Images
		".png":  "image/png",
		".jpg":  "image/jpeg",
		".jpeg": "image/jpeg",
		".gif":  "image/gif",
		".svg":  "image/svg+xml",
		".webp": "image/webp",
		".ico":  "image/x-icon",
		
		// Fonts
		".woff":  "font/woff",
		".woff2": "font/woff2",
		".ttf":   "font/ttf",
		".otf":   "font/otf",
		".eot":   "application/vnd.ms-fontobject",
		
		// Audio/Video
		".mp3":  "audio/mpeg",
		".wav":  "audio/wav",
		".ogg":  "audio/ogg",
		".mp4":  "video/mp4",
		".webm": "video/webm",
		
		// Data
		".csv":  "text/csv",
		".pdf":  "application/pdf",
		".zip":  "application/zip",
		
		// WASM
		".wasm": "application/wasm",
	}
	
	if mimeType, exists := mimeTypes[ext]; exists {
		return mimeType
	}
	
	return "application/octet-stream"
}

// Template methods for common manifest patterns

// CreateInteractiveDocumentTemplate creates a template for interactive documents
func CreateInteractiveDocumentTemplate(title, author string) *ManifestBuilder {
	mb := NewManifestBuilder()
	
	// Set metadata
	mb.CreateDefaultMetadata(title, author)
	
	// Set permissive security policy for interactive content
	policy := &core.SecurityPolicy{
		WASMPermissions: &core.WASMPermissions{
			MemoryLimit:     128 * 1024 * 1024, // 128MB
			AllowedImports:  []string{"env", "wasi_snapshot_preview1"},
			CPUTimeLimit:    15000, // 15 seconds
			AllowNetworking: false,
			AllowFileSystem: false,
		},
		JSPermissions: &core.JSPermissions{
			ExecutionMode: "sandboxed",
			AllowedAPIs:   []string{"canvas", "webgl", "audio"},
			DOMAccess:     "write",
		},
		NetworkPolicy: &core.NetworkPolicy{
			AllowOutbound: false,
			AllowedHosts:  []string{},
			AllowedPorts:  []int{},
		},
		StoragePolicy: &core.StoragePolicy{
			AllowLocalStorage:   true,
			AllowSessionStorage: true,
			AllowIndexedDB:      false,
			AllowCookies:        false,
		},
		ContentSecurityPolicy: "default-src 'self'; script-src 'self' 'unsafe-inline' 'wasm-unsafe-eval'; style-src 'self' 'unsafe-inline';",
		TrustedDomains:        []string{},
	}
	mb.SetSecurityPolicy(policy)
	
	// Enable interactive features
	features := &core.FeatureFlags{
		Animations:    true,
		Interactivity: true,
		Charts:        true,
		Forms:         true,
		Audio:         false,
		Video:         false,
		WebGL:         true,
		WebAssembly:   true,
	}
	mb.SetFeatureFlags(features)
	
	return mb
}

// CreateStaticDocumentTemplate creates a template for static documents
func CreateStaticDocumentTemplate(title, author string) *ManifestBuilder {
	mb := NewManifestBuilder()
	
	// Set metadata
	mb.CreateDefaultMetadata(title, author)
	
	// Set restrictive security policy for static content
	policy := &core.SecurityPolicy{
		WASMPermissions: &core.WASMPermissions{
			// Use minimum allowed values to satisfy validation while effectively disabling WASM
			MemoryLimit:     1024, // 1KB minimum
			AllowedImports:  []string{},
			CPUTimeLimit:    100, // 100ms minimum
			AllowNetworking: false,
			AllowFileSystem: false,
		},
		JSPermissions: &core.JSPermissions{
			ExecutionMode: "none",
			AllowedAPIs:   []string{},
			DOMAccess:     "none",
		},
		NetworkPolicy: &core.NetworkPolicy{
			AllowOutbound: false,
			AllowedHosts:  []string{},
			AllowedPorts:  []int{},
		},
		StoragePolicy: &core.StoragePolicy{
			AllowLocalStorage:   false,
			AllowSessionStorage: false,
			AllowIndexedDB:      false,
			AllowCookies:        false,
		},
		ContentSecurityPolicy: "default-src 'none'; style-src 'self';",
		TrustedDomains:        []string{},
	}
	mb.SetSecurityPolicy(policy)
	
	// Disable interactive features
	features := &core.FeatureFlags{
		Animations:    false,
		Interactivity: false,
		Charts:        false,
		Forms:         false,
		Audio:         false,
		Video:         false,
		WebGL:         false,
		WebAssembly:   false,
	}
	mb.SetFeatureFlags(features)
	
	return mb
}

File: python/src/liv/builder.py
"""
LIV Document Builder for the Python SDK

Provides a fluent API for creating LIV documents programmatically.
"""

from pathlib import Path
from typing import Dict, List, Optional, Any, Union
import mimetypes
import json

from .document import LIVDocument
from .models import (
    DocumentMetadata, SecurityPolicy, WASMPermissions, JSPermissions,
    NetworkPolicy, StoragePolicy, AssetInfo, WASMModuleInfo, FeatureFlags
)
from .exceptions import LIVError, AssetError, ValidationError


class LIVBuilder:
    """Builder class for creating LIV documents with a fluent API."""
    
    def __init__(self):
        """Initialize the builder."""
        self.document = LIVDocument()
        self._reset_to_defaults()
    
    def _reset_to_defaults(self) -> None:
        """Reset document to default values."""
        # Set default metadata
        self.document.metadata = DocumentMetadata(
            title="New LIV Document",
            author="Unknown",
            description="",
            version="1.0",
            language="en"
        )
        
        # Set default security policy
        self.document.security_policy = SecurityPolicy(
            wasm_permissions=WASMPermissions(
                memory_limit=64 * 1024 * 1024,  # 64MB
                cpu_time_limit=5000,  # 5 seconds
                allow_networking=False,
                allow_file_system=False,
                allowed_imports=["env"]
            ),
            js_permissions=JSPermissions(
                execution_mode="sandboxed",
                allowed_apis=["dom"],
                dom_access="read"
            ),
            network_policy=NetworkPolicy(
                allow_outbound=False,
                allowed_hosts=[],
                allowed_ports=[]
            ),
            storage_policy=StoragePolicy(
                allow_local_storage=False,
                allow_session_storage=False,
                allow_indexed_db=False,
                allow_cookies=False
            )
        )
        
        # Set default feature flags
        self.document.feature_flags = FeatureFlags()
        
        # Initialize empty content
        self.document.html_content = ""
        self.document.css_content = ""
        self.document.js_content = ""
        self.document.static_fallback = ""
        
        # Clear assets and modules
        self.document.assets = {}
        self.document.wasm_modules = {}
    
    def set_metadata(self, title: Optional[str] = None, author: Optional[str] = None,
                    description: Optional[str] = None, version: Optional[str] = None,
                    language: Optional[str] = None, keywords: Optional[List[str]] = None,
                    **custom_fields) -> 'LIVBuilder':
        """
        Set document metadata.
        
        Args:
            title: Document title
            author: Document author
            description: Document description
            version: Document version
            language: Document language
            keywords: List of keywords
            **custom_fields: Additional custom metadata fields
            
        Returns:
            Builder instance for chaining
        """
        if not self.document.metadata:
            self.document.metadata = DocumentMetadata(title="", author="")
        
        if title is not None:
            self.document.metadata.title = title
        if author is not None:
            self.document.metadata.author = author
        if description is not None:
            self.document.metadata.description = description
        if version is not None:
            self.document.metadata.version = version
        if language is not None:
            self.document.metadata.language = language
        if keywords is not None:
            self.document.metadata.keywords = keywords
        
        # Add custom fields
        self.document.metadata.custom_fields.update(custom_fields)
        
        return self
    
    def set_content(self, html: Optional[str] = None, css: Optional[str] = None,
                   js: Optional[str] = None, static_fallback: Optional[str] = None) -> 'LIVBuilder':
        """
        Set document content.
        
        Args:
            html: HTML content
            css: CSS content
            js: JavaScript content
            static_fallback: Static fallback HTML
            
        Returns:
            Builder instance for chaining
        """
        if html is not None:
            self.document.html_content = html
        if css is not None:
            self.document.css_content = css
        if js is not None:
            self.document.js_content = js
            # Enable interactivity if JS is provided
            if js and self.document.feature_flags:
                self.document.feature_flags.interactivity = True
        if static_fallback is not None:
            self.document.static_fallback = static_fallback
        
        return self
    
    def set_html(self, html: str) -> 'LIVBuilder':
        """Set HTML content."""
        self.document.html_content = html
        return self
    
    def set_css(self, css: str) -> 'LIVBuilder':
        """Set CSS content."""
        self.document.css_content = css
        # Check for animations in CSS
        if css and self.document.feature_flags:
            if any(keyword in css.lower() for keyword in ['@keyframes', 'animation:', 'transition:']):
                self.document.feature_flags.animations = True
        return self
    
    def set_javascript(self, js: str) -> 'LIVBuilder':
        """Set JavaScript content."""
        self.document.js_content = js
        if js and self.document.feature_flags:
            self.document.feature_flags.interactivity = True
        return self
    
    def set_static_fallback(self, fallback: str) -> 'LIVBuilder':
        """Set static fallback content."""
        self.document.static_fallback = fallback
        return self
    
    def load_content_from_files(self, html_file: Optional[Union[str, Path]] = None,
                               css_file: Optional[Union[str, Path]] = None,
                               js_file: Optional[Union[str, Path]] = None) -> 'LIVBuilder':
        """
        Load content from files.
        
        Args:
            html_file: Path to HTML file
            css_file: Path to CSS file
            js_file: Path to JavaScript file
            
        Returns:
            Builder instance for chaining
        """
        if html_file:
            html_path = Path(html_file)
            if html_path.exists():
                with open(html_path, 'r', encoding='utf-8') as f:
                    self.set_html(f.read())
            else:
                raise LIVError(f"HTML file not found: {html_file}")
        
        if css_file:
            css_path = Path(css_file)
            if css_path.exists():
                with open(css_path, 'r', encoding='utf-8') as f:
                    self.set_css(f.read())
            else:
                raise LIVError(f"CSS file not found: {css_file}")
        
        if js_file:
            js_path = Path(js_file)
            if js_path.exists():
                with open(js_path, 'r', encoding='utf-8') as f:
                    self.set_javascript(f.read())
            else:
                raise LIVError(f"JavaScript file not found: {js_file}")
        
        return self
    
    def set_security_policy(self, policy: SecurityPolicy) -> 'LIVBuilder':
        """Set security policy."""
        self.document.security_policy = policy
        return self
    
    def set_wasm_permissions(self, memory_limit: Optional[int] = None,
                           cpu_time_limit: Optional[int] = None,
                           allow_networking: Optional[bool] = None,
                           allow_file_system: Optional[bool] = None,
                           allowed_imports: Optional[List[str]] = None) -> 'LIVBuilder':
        """
        Set WASM permissions.
        
        Args:
            memory_limit: Memory limit in bytes
            cpu_time_limit: CPU time limit in milliseconds
            allow_networking: Whether to allow networking
            allow_file_system: Whether to allow file system access
            allowed_imports: List of allowed imports
            
        Returns:
            Builder instance for chaining
        """
        if not self.document.security_policy:
            self.document.security_policy = SecurityPolicy()
        
        wasm_perms = self.document.security_policy.wasm_permissions
        
        if memory_limit is not None:
            wasm_perms.memory_limit = memory_limit
        if cpu_time_limit is not None:
            wasm_perms.cpu_time_limit = cpu_time_limit
        if allow_networking is not None:
            wasm_perms.allow_networking = allow_networking
        if allow_file_system is not None:
            wasm_perms.allow_file_system = allow_file_system
        if allowed_imports is not None:
            wasm_perms.allowed_imports = allowed_imports
        
        return self
    
    def set_js_permissions(self, execution_mode: Optional[str] = None,
                          allowed_apis: Optional[List[str]] = None,
                          dom_access: Optional[str] = None) -> 'LIVBuilder':
        """
        Set JavaScript permissions.
        
        Args:
            execution_mode: Execution mode (none, sandboxed, trusted)
            allowed_apis: List of allowed APIs
            dom_access: DOM access level (none, read, write)
            
        Returns:
            Builder instance for chaining
        """
        if not self.document.security_policy:
            self.document.security_policy = SecurityPolicy()
        
        js_perms = self.document.security_policy.js_permissions
        
        if execution_mode is not None:
            js_perms.execution_mode = execution_mode
        if allowed_apis is not None:
            js_perms.allowed_apis = allowed_apis
        if dom_access is not None:
            js_perms.dom_access = dom_access
        
        return self
    
    def add_asset(self, name: str, asset_type: str, 
                 file_path: Optional[Union[str, Path]] = None,
                 data: Optional[bytes] = None,
                 mime_type: Optional[str] = None) -> 'LIVBuilder':
        """
        Add an asset to the document.
        
        Args:
            name: Asset name
            asset_type: Asset type (image, font, data, audio, video)
            file_path: Path to asset file
            data: Asset data (if not loading from file)
            mime_type: MIME type (auto-detected if not provided)
            
        Returns:
            Builder instance for chaining
        """
        if file_path and data:
            raise AssetError("Cannot specify both file_path and data")
        
        if not file_path and not data:
            raise AssetError("Must specify either file_path or data")
        
        # Load from file if path provided
        if file_path:
            path = Path(file_path)
            if not path.exists():
                raise AssetError(f"Asset file not found: {file_path}", asset_path=str(file_path))
            
            with open(path, 'rb') as f:
                data = f.read()
            
            # Auto-detect MIME type if not provided
            if not mime_type:
                mime_type, _ = mimetypes.guess_type(str(path))
        
        # Create asset info
        asset_info = AssetInfo(
            name=name,
            asset_type=asset_type,
            path=Path(file_path) if file_path else None,
            data=data,
            mime_type=mime_type,
            size=len(data) if data else None
        )
        
        self.document.assets[name] = asset_info
        
        # Update feature flags based on asset type
        if self.document.feature_flags:
            if asset_type == "audio":
                self.document.feature_flags.audio = True
            elif asset_type == "video":
                self.document.feature_flags.video = True
        
        return self
    
    def add_image(self, name: str, file_path: Union[str, Path], 
                 mime_type: Optional[str] = None) -> 'LIVBuilder':
        """Add an image asset."""
        return self.add_asset(name, "image", file_path=file_path, mime_type=mime_type)
    
    def add_font(self, name: str, file_path: Union[str, Path],
                mime_type: Optional[str] = None) -> 'LIVBuilder':
        """Add a font asset."""
        return self.add_asset(name, "font", file_path=file_path, mime_type=mime_type)
    
    def add_data(self, name: str, data: Union[str, bytes, Dict[str, Any]],
                mime_type: str = "application/octet-stream") -> 'LIVBuilder':
        """
        Add a data asset.
        
        Args:
            name: Asset name
            data: Data content (string, bytes, or dict for JSON)
            mime_type: MIME type
            
        Returns:
            Builder instance for chaining
        """
        if isinstance(data, dict):
            data_bytes = json.dumps(data, indent=2).encode('utf-8')
            mime_type = "application/json"
        elif isinstance(data, str):
            data_bytes = data.encode('utf-8')
            if mime_type == "application/octet-stream":
                mime_type = "text/plain"
        else:
            data_bytes = data
        
        return self.add_asset(name, "data", data=data_bytes, mime_type=mime_type)
    
    def add_wasm_module(self, name: str, file_path: Optional[Union[str, Path]] = None,
                       data: Optional[bytes] = None, version: str = "1.0",
                       entry_point: str = "main",
                       permissions: Optional[WASMPermissions] = None) -> 'LIVBuilder':
        """
        Add a WASM module to the document.
        
        Args:
            name: Module name
            file_path: Path to WASM file
            data: WASM module data (if not loading from file)
            version: Module version
            entry_point: Entry point function name
            permissions: Module-specific permissions
            
        Returns:
            Builder instance for chaining
        """
        if file_path and data:
            raise LIVError("Cannot specify both file_path and data")
        
        if not file_path and not data:
            raise LIVError("Must specify either file_path or data")
        
        # Load from file if path provided
        if file_path:
            path = Path(file_path)
            if not path.exists():
                raise LIVError(f"WASM file not found: {file_path}")
            
            with open(path, 'rb') as f:
                data = f.read()
        
        # Create module info
        module_info = WASMModuleInfo(
            name=name,
            path=Path(file_path) if file_path else None,
            data=data,
            version=version,
            entry_point=entry_point,
            permissions=permissions
        )
        
        self.document.wasm_modules[name] = module_info
        
        # Enable WASM feature flag
        if self.document.feature_flags:
            self.document.feature_flags.webassembly = True
        
        return self
    
    def enable_features(self, animations: Optional[bool] = None,
                       interactivity: Optional[bool] = None,
                       charts: Optional[bool] = None,
                       forms: Optional[bool] = None,
                       audio: Optional[bool] = None,
                       video: Optional[bool] = None,
                       webgl: Optional[bool] = None,
                       webassembly: Optional[bool] = None) -> 'LIVBuilder':
        """
        Enable or disable document features.
        
        Args:
            animations: Enable animations
            interactivity: Enable interactivity
            charts: Enable charts
            forms: Enable forms
            audio: Enable audio
            video: Enable video
            webgl: Enable WebGL
            webassembly: Enable WebAssembly
            
        Returns:
            Builder instance for chaining
        """
        if not self.document.feature_flags:
            self.document.feature_flags = FeatureFlags()
        
        flags = self.document.feature_flags
        
        if animations is not None:
            flags.animations = animations
        if interactivity is not None:
            flags.interactivity = interactivity
        if charts is not None:
            flags.charts = charts
        if forms is not None:
            flags.forms = forms
        if audio is not None:
            flags.audio = audio
        if video is not None:
            flags.video = video
        if webgl is not None:
            flags.webgl = webgl
        if webassembly is not None:
            flags.webassembly = webassembly
        
        return self
    
    def validate(self) -> List[str]:
        """
        Validate the current document configuration.
        
        Returns:
            List of validation errors (empty if valid)
        """
        errors = []
        
        # Check required metadata
        if not self.document.metadata:
            errors.append("Document metadata is required")
        else:
            if not self.document.metadata.title:
                errors.append("Document title is required")
            if not self.document.metadata.author:
                errors.append("Document author is required")
        
        # Check content
        if not self.document.html_content and not self.document.static_fallback:
            errors.append("Document must have HTML content or static fallback")
        
        # Check WASM modules have valid data
        for name, module in self.document.wasm_modules.items():
            if not module.data and not (module.path and module.path.exists()):
                errors.append(f"WASM module '{name}' has no data or invalid path")
        
        # Check assets have valid data
        for name, asset in self.document.assets.items():
            if not asset.data and not (asset.path and asset.path.exists()):
                errors.append(f"Asset '{name}' has no data or invalid path")
        
        return errors
    
    def build(self) -> LIVDocument:
        """
        Build and return the LIV document.
        
        Returns:
            Completed LIVDocument instance
            
        Raises:
            ValidationError: If document validation fails
        """
        # Validate before building
        errors = self.validate()
        if errors:
            raise ValidationError("Document validation failed", errors=errors)
        
        return self.document
    
    def build_and_save(self, output_path: Union[str, Path],
                      sign: bool = False, key_path: Optional[Union[str, Path]] = None) -> LIVDocument:
        """
        Build document and save to file.
        
        Args:
            output_path: Output file path
            sign: Whether to sign the document
            key_path: Path to signing key
            
        Returns:
            Completed LIVDocument instance
        """
        document = self.build()
        document.save(output_path, sign=sign, key_path=key_path)
        return document
    
    def reset(self) -> 'LIVBuilder':
        """Reset builder to initial state."""
        self.document = LIVDocument()
        self._reset_to_defaults()
        return self
    
    def clone(self) -> 'LIVBuilder':
        """Create a copy of the current builder."""
        new_builder = LIVBuilder()
        
        # Copy metadata
        if self.document.metadata:
            new_builder.document.metadata = DocumentMetadata.from_dict(
                self.document.metadata.to_dict()
            )
        
        # Copy security policy
        if self.document.security_policy:
            new_builder.document.security_policy = SecurityPolicy.from_dict(
                self.document.security_policy.to_dict()
            )
        
        # Copy feature flags
        if self.document.feature_flags:
            new_builder.document.feature_flags = FeatureFlags.from_dict(
                self.document.feature_flags.to_dict()
            )
        
        # Copy content
        new_builder.document.html_content = self.document.html_content
        new_builder.document.css_content = self.document.css_content
        new_builder.document.js_content = self.document.js_content
        new_builder.document.static_fallback = self.document.static_fallback
        
        # Copy assets and modules (shallow copy)
        new_builder.document.assets = self.document.assets.copy()
        new_builder.document.wasm_modules = self.document.wasm_modules.copy()
        
        return new_builder


# Convenience functions
def create_simple_document(title: str, author: str, html_content: str,
                          css_content: str = "", js_content: str = "") -> LIVDocument:
    """Create a simple LIV document with basic content."""
    builder = LIVBuilder()
    return (builder
            .set_metadata(title=title, author=author)
            .set_content(html=html_content, css=css_content, js=js_content)
            .build())


def create_from_html_file(html_file: Union[str, Path], title: str, author: str,
                         css_file: Optional[Union[str, Path]] = None,
                         js_file: Optional[Union[str, Path]] = None) -> LIVDocument:
    """Create a LIV document from HTML file."""
    builder = LIVBuilder()
    return (builder
            .set_metadata(title=title, author=author)
            .load_content_from_files(html_file=html_file, css_file=css_file, js_file=js_file)
            .build())


__all__ = [
    "LIVBuilder",
    "create_simple_document",
    "create_from_html_file",
]